#Libraries

In [1]:
import pandas as pd
import numpy as np

# Download the CSV from Kaggle and upload to Colab

In [2]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("blastchar/telco-customer-churn")

print("Path to dataset files:", path)

Using Colab cache for faster access to the 'telco-customer-churn' dataset.
Path to dataset files: /kaggle/input/telco-customer-churn


In [3]:
df = pd.read_csv(path + '/WA_Fn-UseC_-Telco-Customer-Churn.csv')
df

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,...,Yes,Yes,Yes,Yes,One year,Yes,Mailed check,84.80,1990.5,No
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,...,Yes,No,Yes,Yes,One year,Yes,Credit card (automatic),103.20,7362.9,No
7040,4801-JZAZL,Female,0,Yes,Yes,11,No,No phone service,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.60,346.45,No
7041,8361-LTMKD,Male,1,Yes,No,4,Yes,Yes,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Mailed check,74.40,306.6,Yes


In [4]:
df.shape

(7043, 21)

In [5]:
df.dtypes

,0
customerID,object
gender,object
SeniorCitizen,int64
Partner,object
Dependents,object
tenure,int64
PhoneService,object
MultipleLines,object
InternetService,object
OnlineSecurity,object


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [7]:
df.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


In [8]:
df.describe(include='object')

,customerID,gender,Partner,Dependents,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,TotalCharges,Churn
count,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043,7043
unique,7043,2,2,2,2,3,3,3,3,3,3,3,3,3,2,4,6531,2
top,3186-AJIEK,Male,No,No,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,,No
freq,1,3555,3641,4933,6361,3390,3096,3498,3088,3095,3473,2810,2785,3875,4171,2365,11,5174


In [9]:
# Step 6: Classify columns as Numerical vs Categorical
numerical_cols   = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_cols = df.select_dtypes(include='object').columns.tolist()

print(f"Numerical   ({len(numerical_cols)}): {numerical_cols}")
print(f"Categorical ({len(categorical_cols)}): {categorical_cols}")

Numerical   (3): ['SeniorCitizen', 'tenure', 'MonthlyCharges']
Categorical (18): ['customerID', 'gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'TotalCharges', 'Churn']


#Identify and Handle Missing Values

##count NaN values per column

In [10]:
df.isnull().sum()

,0
customerID,0
gender,0
SeniorCitizen,0
Partner,0
Dependents,0
tenure,0
PhoneService,0
MultipleLines,0
InternetService,0
OnlineSecurity,0


## Check for hidden blanks

In [11]:
df[df['TotalCharges'] == ' ']

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,Yes,Bank transfer (automatic),52.55,,No
753,3115-CZMZD,Male,0,No,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.25,,No
936,5709-LVOEQ,Female,0,Yes,Yes,0,Yes,No,DSL,Yes,...,Yes,No,Yes,Yes,Two year,No,Mailed check,80.85,,No
1082,4367-NUYAO,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.75,,No
1340,1371-DWPAZ,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,No,Credit card (automatic),56.05,,No
3331,7644-OMVMY,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,19.85,,No
3826,3213-VVOLG,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.35,,No
4380,2520-SGTTA,Female,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.00,,No
5218,2923-ARZLG,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,One year,Yes,Mailed check,19.70,,No
6670,4075-WKNIU,Female,0,Yes,Yes,0,Yes,Yes,DSL,No,...,Yes,Yes,Yes,No,Two year,No,Mailed check,73.35,,No


## Convert TotalCharges

In [12]:
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'],errors='coerce')

##Create a missing indicator before imputing

### Missingness Classification

**TotalCharges (11 ta blank → NaN):**
Barcha 11 ta yetishmayotgan qiymat tenure=0 bo'lgan
mijozlarga tegishli — ular hali to'lov olmagan.
Bu **MNAR (Missing Not At Random)** — missingness
boshqa o'zgaruvchi (tenure) qiymatiga bog'liq,
tasodifiy emas.

In [13]:
df['TotalCharges_missing'] = df['TotalCharges'].isnull().astype(int)

## Impute numerical columns with median

In [14]:
df['TotalCharges'] = df['TotalCharges'].fillna(df['TotalCharges'].median())

In [15]:
# Step 7: Check & impute categorical columns with mode
for col in df.select_dtypes(include='object').columns:
    if df[col].isnull().sum() > 0:
        df[col] = df[col].fillna(df[col].mode()[0])
        print(f"Imputed '{col}' with mode: {df[col].mode()[0]}")

print("Categorical missing check done.")

Categorical missing check done.


# Standardize Inconsistent Categorical Entries

In [16]:
df['gender'].value_counts()

,count
gender,
Male,3555
Female,3488


In [17]:
# To get unique values, you typically apply .unique() to a specific column (Series).
# For example, to see unique values in the 'gender' column:
df['gender'].unique()

array(['Female', 'Male'], dtype=object)

To get unique values for all columns, we can loop through each column and apply the `.unique()` method.

In [18]:
for column in df.columns:
    print(f"Unique values in column '{column}':\n{df[column].unique()}\n")

Unique values in column 'customerID':
['7590-VHVEG' '5575-GNVDE' '3668-QPYBK' ... '4801-JZAZL' '8361-LTMKD'
 '3186-AJIEK']

Unique values in column 'gender':
['Female' 'Male']

Unique values in column 'SeniorCitizen':
[0 1]

Unique values in column 'Partner':
['Yes' 'No']

Unique values in column 'Dependents':
['No' 'Yes']

Unique values in column 'tenure':
[ 1 34  2 45  8 22 10 28 62 13 16 58 49 25 69 52 71 21 12 30 47 72 17 27
  5 46 11 70 63 43 15 60 18 66  9  3 31 50 64 56  7 42 35 48 29 65 38 68
 32 55 37 36 41  6  4 33 67 23 57 61 14 20 53 40 59 24 44 19 54 51 26  0
 39]

Unique values in column 'PhoneService':
['No' 'Yes']

Unique values in column 'MultipleLines':
['No phone service' 'No' 'Yes']

Unique values in column 'InternetService':
['DSL' 'Fiber optic' 'No']

Unique values in column 'OnlineSecurity':
['No' 'Yes' 'No internet service']

Unique values in column 'OnlineBackup':
['Yes' 'No' 'No internet service']

Unique values in column 'DeviceProtection':
['No' 'Yes' 'No in

To get the frequency count of unique values for all columns, we can loop through each column and apply the `.value_counts()` method.

In [19]:
for column in df.columns:
    print(f"Value counts for column '{column}':\n{df[column].value_counts()}\n")

Value counts for column 'customerID':
customerID
3186-AJIEK    1
7590-VHVEG    1
5575-GNVDE    1
4501-VCPFK    1
6075-SLNIL    1
             ..
1452-KIOVK    1
6713-OKOMC    1
7892-POOKP    1
6388-TABGU    1
9763-GRSKD    1
Name: count, Length: 7043, dtype: int64

Value counts for column 'gender':
gender
Male      3555
Female    3488
Name: count, dtype: int64

Value counts for column 'SeniorCitizen':
SeniorCitizen
0    5901
1    1142
Name: count, dtype: int64

Value counts for column 'Partner':
Partner
No     3641
Yes    3402
Name: count, dtype: int64

Value counts for column 'Dependents':
Dependents
No     4933
Yes    2110
Name: count, dtype: int64

Value counts for column 'tenure':
tenure
1     613
72    362
2     238
3     200
4     176
     ... 
28     57
39     56
44     51
36     50
0      11
Name: count, Length: 73, dtype: int64

Value counts for column 'PhoneService':
PhoneService
Yes    6361
No      682
Name: count, dtype: int64

Value counts for column 'MultipleLines':
Multi

From the `value_counts()` output, we can observe that columns like `MultipleLines`, `OnlineSecurity`, `OnlineBackup`, `DeviceProtection`, `TechSupport`, `StreamingTV`, and `StreamingMovies` contain 'No internet service' as a category. This can be standardized to 'No' to represent the absence of that particular service.

In [20]:
# Replace 'No internet service' with 'No' for relevant columns
for col in ['MultipleLines', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
            'TechSupport', 'StreamingTV', 'StreamingMovies']:
    if 'No internet service' in df[col].unique():
        df[col] = df[col].replace('No internet service', 'No')

# Also standardize 'No phone service' to 'No' in 'MultipleLines'
if 'No phone service' in df['MultipleLines'].unique():
    df['MultipleLines'] = df['MultipleLines'].replace('No phone service', 'No')

# Verify the changes by checking unique values again
print("Unique values after standardization:")
for col in ['MultipleLines', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
            'TechSupport', 'StreamingTV', 'StreamingMovies']:
    print(f"'{col}': {df[col].unique()}")

Unique values after standardization:
'MultipleLines': ['No' 'Yes']
'OnlineSecurity': ['No' 'Yes']
'OnlineBackup': ['Yes' 'No']
'DeviceProtection': ['No' 'Yes']
'TechSupport': ['No' 'Yes']
'StreamingTV': ['No' 'Yes']
'StreamingMovies': ['No' 'Yes']


To further standardize categorical features, it's often useful to convert all string entries to lowercase. This ensures that variations in casing (e.g., 'Yes' vs. 'yes') are treated as the same category.

In [21]:
# Step 3: Strip leading/trailing whitespace
for col in df.select_dtypes(include='object').columns:
    df[col] = df[col].str.strip()

print("Whitespace stripped from all object columns.")

Whitespace stripped from all object columns.


In [22]:
# Convert all object type columns to lowercase
for col in df.select_dtypes(include='object').columns:
    df[col] = df[col].str.lower()

# Display unique values of a sample column to verify (e.g., 'gender')
print("Unique values in 'gender' after lowercasing:", df['gender'].unique())

Unique values in 'gender' after lowercasing: ['female' 'male']


# Convert Data Types and Extract Temporal Features

In [23]:
df['TotalCharges']

,TotalCharges
0,29.85
1,1889.50
2,108.15
3,1840.75
4,151.65
...,...
7038,1990.50
7039,7362.90
7040,346.45
7041,306.60


In [24]:
df

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,TotalCharges_missing
0,7590-vhveg,female,0,yes,no,1,no,no,dsl,no,...,no,no,no,month-to-month,yes,electronic check,29.85,29.85,no,0
1,5575-gnvde,male,0,no,no,34,yes,no,dsl,yes,...,no,no,no,one year,no,mailed check,56.95,1889.50,no,0
2,3668-qpybk,male,0,no,no,2,yes,no,dsl,yes,...,no,no,no,month-to-month,yes,mailed check,53.85,108.15,yes,0
3,7795-cfocw,male,0,no,no,45,no,no,dsl,yes,...,yes,no,no,one year,no,bank transfer (automatic),42.30,1840.75,no,0
4,9237-hqitu,female,0,no,no,2,yes,no,fiber optic,no,...,no,no,no,month-to-month,yes,electronic check,70.70,151.65,yes,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7038,6840-resvb,male,0,yes,yes,24,yes,yes,dsl,yes,...,yes,yes,yes,one year,yes,mailed check,84.80,1990.50,no,0
7039,2234-xaduh,female,0,yes,yes,72,yes,yes,fiber optic,no,...,no,yes,yes,one year,yes,credit card (automatic),103.20,7362.90,no,0
7040,4801-jzazl,female,0,yes,yes,11,no,no,dsl,yes,...,no,no,no,month-to-month,yes,electronic check,29.60,346.45,no,0
7041,8361-ltmkd,male,1,yes,no,4,yes,yes,fiber optic,no,...,no,no,no,month-to-month,yes,mailed check,74.40,306.60,yes,0


In [25]:
# Step 1: Confirm TotalCharges is float64
print("TotalCharges dtype:", df['TotalCharges'].dtype)
assert df['TotalCharges'].dtype == 'float64', "TotalCharges is NOT float64!"

TotalCharges dtype: float64


In [26]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn',
       'TotalCharges_missing'],
      dtype='object')

In [27]:
df['Contract']

,Contract
0,month-to-month
1,one year
2,month-to-month
3,one year
4,month-to-month
...,...
7038,one year
7039,one year
7040,month-to-month
7041,month-to-month


In [28]:
df['tenure']

,tenure
0,1
1,34
2,2
3,45
4,2
...,...
7038,24
7039,72
7040,11
7041,4


In [29]:
df['tenure_years'] = df['tenure'] / 12

In [30]:
df['tenure_years']

,tenure_years
0,0.083333
1,2.833333
2,0.166667
3,3.750000
4,0.166667
...,...
7038,2.000000
7039,6.000000
7040,0.916667
7041,0.333333


In [31]:
df.dtypes

,0
customerID,object
gender,object
SeniorCitizen,int64
Partner,object
Dependents,object
tenure,int64
PhoneService,object
MultipleLines,object
InternetService,object
OnlineSecurity,object


In [32]:
# Step 2: Check for date columns
date_cols = [col for col in df.columns if 'date' in col.lower()]
if date_cols:
    for col in date_cols:
        df[col] = pd.to_datetime(df[col])
        print(f"Converted '{col}' to datetime.")
else:
    print("No date columns found — no conversion needed.")

No date columns found — no conversion needed.


#Detect and Remove Duplicates

In [33]:
df.duplicated().sum()

np.int64(0)

In [34]:
df[df.duplicated(keep=False)]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,TotalCharges_missing,tenure_years


In [35]:
rows_before = len(df)
df = df.drop_duplicates()
rows_after = len(df)
print(f"Rows before: {rows_before} | After: {rows_after} | Dropped: {rows_before - rows_after}")

Rows before: 7043 | After: 7043 | Dropped: 0


# Identify Unrealistic Values

In [36]:
df[df['MonthlyCharges'] < 0]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn,TotalCharges_missing,tenure_years


In [37]:
df['tenure'].describe()

,tenure
count,7043.000000
mean,32.371149
std,24.559481
min,0.000000
25%,9.000000
50%,29.000000
75%,55.000000
max,72.000000


In [38]:
# Check TotalCharges for customers with tenure = 0
print("Customers with tenure = 0 and their TotalCharges:")
df[df['tenure'] == 0][['customerID', 'tenure', 'TotalCharges']]

Customers with tenure = 0 and their TotalCharges:


,customerID,tenure,TotalCharges
488,4472-lvygi,0,1397.475
753,3115-czmzd,0,1397.475
936,5709-lvoeq,0,1397.475
1082,4367-nuyao,0,1397.475
1340,1371-dwpaz,0,1397.475
3331,7644-omvmy,0,1397.475
3826,3213-vvolg,0,1397.475
4380,2520-sgtta,0,1397.475
5218,2923-arzlg,0,1397.475
6670,4075-wkniu,0,1397.475


### Unrealistic Values and Their Handling

1.  **`TotalCharges` Blank Strings:**
    *   **Detection:** Initially, `TotalCharges` was of object type, and `df[df['TotalCharges'] == ' ']` identified 11 rows where `TotalCharges` contained blank strings.
    *   **Handling:** These blank strings were converted to `NaN` during the `pd.to_numeric(df['TotalCharges'], errors='coerce')` step. A new indicator column, `TotalCharges_missing`, was created to mark these rows. Finally, the `NaN` values in `TotalCharges` were imputed with the median of the column.

2.  **`tenure = 0` with Non-Zero `TotalCharges`:**
    *   **Detection:** After imputation, it was observed that customers with a `tenure` of 0 months (meaning they just joined) still had a `TotalCharges` value equal to the median (1397.475). This is unrealistic, as a customer with 0 tenure should theoretically have 0 total charges.
    *   **Handling:** For the purpose of this analysis, these values were **left as is** following the median imputation step. It's noted as an inconsistency that might require further domain-specific handling (e.g., setting `TotalCharges` to 0 for these specific cases if `MonthlyCharges` was also 0, or if they represent initial setup fees). However, given the median imputation was already applied to handle missing values, and these were considered 'missing' (blanks) initially, this was the direct consequence of that imputation strategy.

3.  **Negative `MonthlyCharges`:**
    *   **Detection:** A check using `df[df['MonthlyCharges'] < 0]` was performed.
    *   **Handling:** No unrealistic negative values were found in the `MonthlyCharges` column.

4.  **Duplicate Rows:**
    *   **Detection:** Checks using `df.duplicated().sum()` and `df[df.duplicated(keep=False)]` were performed.
    *   **Handling:** No duplicate rows were found in the dataset, so no removal was necessary.

In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   customerID            7043 non-null   object 
 1   gender                7043 non-null   object 
 2   SeniorCitizen         7043 non-null   int64  
 3   Partner               7043 non-null   object 
 4   Dependents            7043 non-null   object 
 5   tenure                7043 non-null   int64  
 6   PhoneService          7043 non-null   object 
 7   MultipleLines         7043 non-null   object 
 8   InternetService       7043 non-null   object 
 9   OnlineSecurity        7043 non-null   object 
 10  OnlineBackup          7043 non-null   object 
 11  DeviceProtection      7043 non-null   object 
 12  TechSupport           7043 non-null   object 
 13  StreamingTV           7043 non-null   object 
 14  StreamingMovies       7043 non-null   object 
 15  Contract             

In [40]:
df.to_csv('telco_cleaned.csv', index=False)
print("DataFrame saved to 'telco_cleaned.csv'")

DataFrame saved to 'telco_cleaned.csv'


## Written Justification

### 1. Missingness Classification
TotalCharges — MNAR. tenure=0 bo'lgan 11 mijoz...

### 2. Imputation Choices
Median tanlandi, chunki TotalCharges right-skewed...

### 3. Rows/Columns Dropped
- Duplicate rows: 0 ta o'chirildi
- Qo'shilgan ustunlar: TotalCharges_missing, tenure_years